sentinel-extension/sentinel-datasource-opensergo/src/main/java/com/alibaba/csp/sentinel/datasource/OpenSergoDataSourceGroup.java

```diff
@@ -0,0 +1,299 @@
+     * @param app       appName to subscribe
+     */
+    public OpenSergoDataSourceGroup(String host, int port, String namespace, String app) {
+        AssertUtil.notEmpty(namespace, "namespace cannot be empty");
+        AssertUtil.notEmpty(app, "app cannot be empty");
+        this.openSergoClient = new OpenSergoClient(host, port);
+        this.namespace = namespace;
+        this.app = app;
+        this.ruleAggregator = new OpenSergoRuleAggregator(dataSourceMap);
+
+        initializeDataSourceMap();
+    }
+
+    private void initializeDataSourceMap() {
+        dataSourceMap.put(OpenSergoSentinelConstants.KIND_FLOW_RULE, new DynamicSentinelProperty<List<FlowRule>>());
+        dataSourceMap.put(OpenSergoSentinelConstants.KIND_CIRCUIT_BREAKER_RULE,
+            new DynamicSentinelProperty<List<DegradeRule>>());
+    }
+
+    public void start() throws Exception {
+        if (started.compareAndSet(false, true)) {
+            openSergoClient.start();
+
+            prepareForBaseSubscription();
+        }
+    }
+
+    public void close() throws Exception {
+        if (started.compareAndSet(true, false)) {
+            for (Entry<ConfigKind, SubscribeKey> e : subscribeKeyMap.entrySet()) {
+                openSergoClient.unsubscribeConfig(e.getValue());
+                RecordLog.info("Unsubscribing OpenSergo config for target: {}", e.getValue());
+            }
+
+            openSergoClient.close();
+        }
+    }
+
+    private void prepareForBaseSubscription() {
+        SubscribeKey key = new SubscribeKey(namespace, app, ConfigKind.FAULT_TOLERANCE_RULE);
+
+        openSergoClient.subscribeConfig(key, new OpenSergoFaultToleranceRuleSubscriber(ruleAggregator));
+        subscribeKeyMap.put(ConfigKind.FAULT_TOLERANCE_RULE, key);
+
+        RecordLog.info("Subscribing OpenSergo base fault-tolerance rules for target <{}, {}>", namespace, app);
+    }
+
+    public boolean hasSubscribedFor(String sentinelRuleKind) {
+        if (sentinelRuleKind == null) {
+            return false;
+        }
+        return sentinelRuleSubscribeMap.getOrDefault(sentinelRuleKind, false);
+    }
+
+    /**
+     * <p>Subscribe Sentinel flow rules from OpenSergo control plane.</p>
+     *
+     * @return {@link SentinelProperty} of Sentinel flow rules
+     */
+    public synchronized SentinelProperty<List<FlowRule>> subscribeFlowRules() {
+        if (!started.get()) {
+            throw new IllegalStateException("OpenSergo data-source not started");
+        }
+
+        boolean subscribed = hasSubscribedFor(OpenSergoSentinelConstants.KIND_FLOW_RULE);
+        SentinelProperty<List<FlowRule>> property = dataSourceMap.get(OpenSergoSentinelConstants.KIND_FLOW_RULE);
+        if (subscribed) {
+            return property;
+        }
+
+        OpenSergoSentinelFlowRuleSubscriber subscriber = new OpenSergoSentinelFlowRuleSubscriber(ruleAggregator);
+
+        SubscribeKey rlsKey = new SubscribeKey(namespace, app, ConfigKind.RATE_LIMIT_STRATEGY);
```

Comment:
Does the key here only provide the configuration of the current limiting strategy? Maybe there are some additional keys, such as the key of the downgrade policy